src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/evolutionaryAlgorithm/TerminateByDistancePercAvg.java

```diff
@@ -0,0 +1,53 @@
+package se.liu.ida.hefquin.engine.queryproc.impl.optimizer.evolutionaryAlgorithm;
+
+import java.util.List;
+
+/**
+ * Distance-based termination criterion:
+ *
+ * Termination is triggered when the relative distance between the average cost of plans
+ * in the current generation and in the previous generation has not exceeded
+ * a specified distance threshold for a number of generations.
+ */
+
+public class TerminateByDistancePercAvg implements TerminationCriterion{
+    protected final double percAvgThreshold;
+    protected final int nrGenerations;
+
+    public TerminateByDistancePercAvg( final double percAvgThreshold, final int nrGenerations ) {
+        this.percAvgThreshold = percAvgThreshold;
+        this.nrGenerations = nrGenerations;
+    }
+
+    @Override
+    public boolean readyToTerminate( final Generation currentGeneration, final List<Generation> allPreviousGenerations ) {
+        final int previousGenerationNr = allPreviousGenerations.size();
+        if ( previousGenerationNr < nrGenerations ) {
+            return false;
+        }
+
+        double avgCurrentCost = currentGeneration.avgCost;
+        double avgPreviousCost = allPreviousGenerations.get( previousGenerationNr-1 ).avgCost;
+        double relDistance = ( avgPreviousCost - avgCurrentCost ) / avgPreviousCost;
+
+        if ( relDistance > percAvgThreshold ) {
+            return false;
+        }
```

Comment:
I think you can remove these lines and adjust the loop in a way such it takes care of this first step as well. That way, you are not repeating the same code twice.

The same comment may apply to the other termination criteria as well.